contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -78,16 +78,158 @@ pod placement, volume binding and provisioning should be more tightly coupled wi
+  kinds of volume plugins, each with their own topology labeling scheme.
+
+
+### Example PVs with NodeAffinity
+#### Local Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: local-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  local:
+    path: /mnt/disks/ssd1
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+          - node-1
+```
+
+#### Zonal Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: zonal-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+```
+
+#### Multi-Zonal Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: multi-zonal-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+          - us-central1-b
+```
 
-## Design
+### Default Specification
+Existing admission controllers and dynamic provisioners for zonal volumes
+will be updated to specify PV NodeAffinity in addition to the existing zone
+and region labels.  This will handle newly created PV objects.
+
+Existing PV objects will have to be upgraded to use the new NodeAffinity field.
+This does not have to occur instantaneously, and can be updated within the
+deprecation period.
+
+TODO: This can be done through one of the following methods:
+- Manual updates/scripts
+- cluster/update-storage-objects.sh?
+- A new PV update controller
+
+### Bound PVC Enforcement
+For PVCs that are already bound to a PV with NodeAffinity, enforcement is
+simple and will be done at two places:
+* Scheduler predicate: if a Pod references a PVC that is bound to a PV with
+NodeAffinity, the predicate will evaluate the `Required` NodeSelector against
+the Node's labels to filter the nodes that the Pod can be schedule to.  The
+existing VolumeZone scheduling predicate will coexist with this new predicate
+for several releases until PV NodeAffinity becomes GA and we can deprecate the
+old predicate.
+* Kubelet: PV NodeAffinity is verified against the Node when mounting PVs.
+
+### Unbound PVC Binding
+As mentioned in the problem statement, volume binding occurs without any input
+about a Pod's scheduling constraints.  To fix this, we will delay volume binding
+and provisioning until a Pod is created.  This behavior change will be opt-in as a
+new StorageClass parameter.
```

Comment:
link to some other discussion about the late-provisioning?

Reply:
The detailed discussion is in this same document further down.  Are there any missing details that should be added?